### PR TITLE
osd: Refinements to mclock built-in profiles implementation.

### PR DESCRIPTION
**Pull request by sseshasa**
Add refinements to the mclock built-in profiles implementation. 

- Introduce a simple structure to hold mclock parameters associated
with the active profile (reservation, weight & limit) for each type of 
client class namely "client", "background_recovery" & 
"background_best_effort".
- Set "high_client_ops" as the default profile. 
- Fix the OpSchedulerItem cost scaling calculation.

Another important point to note is that for "high_recovery_ops"
and "balanced" profiles, multiple tests were run with different combinations
of reservation, weights and limits. Due to the current cost model, recovery
ops having high item costs (20MiB and beyond) experience higher latency
even if the mclock limit is set to the max_osd_capacity. Setting this to very
high values (for e.g. 999999) leads to other clients being starved which is
not desirable.

Due to the above, recovery throughput suffers. In order to address this,
the recovery  limits are set to 2x and 1.5x the max_osd_capacity. This
appears to resolve the latency issue for recovery ops and still allow
other mclock clients to have the desired throughputs.

The above issue will get resolved once we start using actual benchmark
data instead of the cost model currently employed.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/osd/scheduler/mClockScheduler.cc [hunk 1]
```diff
@@ -425,18 +405,18 @@ void mClockScheduler::dump(ceph::Formatter &f) const
...
 
   // TODO: move this check into OpSchedulerItem, handle backwards compat
-  if (op_scheduler_class::immediate == item.get_scheduler_class()) {
+  if (op_scheduler_class::immediate == id.class_id) {
     immediate.push_front(std::move(item));
   } else {
+    auto op_type = item.get_op_type();
+    int cost = client_scaled_cost_infos[op_type];
+    // Re-calculate the scaled cost for the client if the item cost changed
+    if (maybe_update_client_cost_info(op_type, item.get_cost())) {
```
Comment (jdurgin) [inline src/osd/scheduler/mClockScheduler.cc:416]: have you tried always calling calc_scaled_cost? I wonder if part of the effect you're seeing with recovery is from using cached cost values that override a range of differences

File: src/osd/scheduler/mClockScheduler.cc [hunk 2]
```diff
@@ -383,31 +376,18 @@ int mClockScheduler::calc_scaled_cost(op_type_t op_type, int cost)
...
-  // Note: Cap the scaling of item cost to ~4MiB as the tag increments
-  // beyond this point are too long causing performance issues. This may
-  // need to be in place until benchmark data is available or a better
-  // scaling model can be put in place. This is a TODO.
-  if (new_cost >= capped_item_cost) {
-    new_cost = capped_item_cost;
-  }
-
   bool cost_changed =
     ((new_cost >= (cur_cost << 1)) || (cur_cost >= (new_cost << 1)));
```
Comment (jdurgin) [inline src/osd/scheduler/mClockScheduler.cc:387]: Does this caching have any effect on performance? If not, let's get rid of it - it makes it much simpler to understand.

With a range of costs treated the same, the actual cost is dependent on the I/O sequence and could have unintended effects if e.g. the first i/o is 51% the size of subsequent I/Os

File: src/osd/scheduler/mClockScheduler.cc [hunk 3]
```diff
@@ -139,45 +137,148 @@ void mClockScheduler::set_osd_mclock_cost_per_io()
...
-    return default_allocation;
-  }
+void mClockScheduler::set_balanced_profile_allocations()
+{
+  // Client Allocation:
+  //   reservation: 40% | weight: 1 | limit: 100% |
+  // Background Recovery Allocation:
+  //   reservation: 40% | weight: 1 | limit: 150% |
+  // Background Best Effort Allocation:
+  //   reservation: 20% | weight: 1 | limit: max |
```
Comment (jdurgin) [inline src/osd/scheduler/mClockScheduler.cc:157]: I think we'd want a limit on best effort for all these profiles

File: src/osd/scheduler/mClockScheduler.cc (hunk 1 above)
Comment (sseshasa) [reply src/osd/scheduler/mClockScheduler.cc]: I tried calling the calc_scaled_cost always() with the latest changes (with pre-calculated osd_max_bandwidth) and pretty
much saw the same performance numbers. I do see the average client latency when using the "high_client_ops" profile to
be marginally better by a 10th of a millisecond in one of the runs. Therefore, it does look like using the cached cost values
doesn't help in the performance aspect. I removed the cached logic and always calculate cost now. I will re-post the
performance numbers with the latest change (which I suspect won't change much).

File: src/osd/scheduler/mClockScheduler.cc (hunk 2 above)
Comment (sseshasa) [reply src/osd/scheduler/mClockScheduler.cc]: I have removed the caching as per the previous comment.

File: src/osd/scheduler/mClockScheduler.cc (hunk 3 above)
Comment (sseshasa) [reply src/osd/scheduler/mClockScheduler.cc]: Sure, I wasn't sure about the limits for best effort clients. For now I have provided the following limits,
balanced - 100% of max_osd_capacity
high_client_ops  and high_recovery_ops - 50% of max_osd_capacity.

I think a new profile catering to the best effort clients may be warranted. At this point I am not sure how much allocation
is necessary for clients like scrubbing, snaptrim and pg_delete and how to test them.

**Review by jdurgin — APPROVED** [review]

**Review by sunnyku — APPROVED** [review]
LGTM

**Comment by sseshasa** [comment]
Pulpito run details can be found here:
https://pulpito.ceph.com/sseshasa-2021-02-02_08:37:37-rados-wip-sseshasa-testing-2021-02-02-1136-distro-basic-smithi/

**Comment by sseshasa** [comment]
Ran teuthology job with tweaks to the high_client_ops profile that gives more allocation and limits to the best_effort_clients (scrub, pg_deletes etc.)
https://pulpito.ceph.com/sseshasa-2021-02-14_18:00:38-rados-wip-dmclock-fix-high-client-profile-distro-basic-smithi/

**Comment by sseshasa** [comment]
retest this please

**Comment by sseshasa** [comment]
retest this please

**Comment by sseshasa** [comment]
retest this please

**Comment by sseshasa** [comment]
Link to the spreadsheet containing test results with mclock profiles:
https://docs.google.com/spreadsheets/d/13roSdcOO9Nx_I-Y9xFSLrRb_VBP2bzKLiLZ1gPOkTvo/edit?usp=sharing

Latest pulpito run:
https://pulpito.ceph.com/sseshasa-2021-02-23_04:18:11-rados-wip-sseshasa-testing-2021-02-22-1136-distro-basic-smithi/

**Review by jdurgin — APPROVED** [review]
looks good!

**Comment by neha-ojha** [comment]
https://pulpito.ceph.com/sseshasa-2021-02-23_04:18:11-rados-wip-sseshasa-testing-2021-02-22-1136-distro-basic-smithi/ - no related failures

**Pull request merged**